contracts/utils/helpers.cairo

```diff
@@ -0,0 +1,386 @@
+%lang starknet
+
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.cairo_builtins import (HashBuiltin,
+    BitwiseBuiltin)
+from starkware.cairo.common.math import (assert_not_zero,
+    assert_le_felt, assert_not_equal, unsigned_div_rem, split_felt)
+from starkware.starknet.common.syscalls import get_caller_address
+from starkware.cairo.common.memcpy import memcpy
+from starkware.cairo.common.uint256 import Uint256
+
+from contracts.utils.life_rules import evaluate_rounds
+from contracts.utils.packing import (unpack_game, pack_game)
+from contracts.utils.constants import (DIM, INFINITE_GAME_GENESIS)
+
+from openzeppelin.token.erc20.library import ERC20
+
+##### Storage #####
+# Records game history on chain.
+# Game id is to ensure game uniqueness.
+@storage_var
+func stored_game(
+        game_id : felt,
+        generation : felt,
+    ) -> (
+        owner_and_state : (felt, felt)
+    ):
+end
+
+# Lets you find the latest state of a given game.
+@storage_var
+func current_generation(
+        game_id : felt
+    ) -> (
+        game_generation : felt
+    ):
+end
+
+##### Events #####
+@event
+func game_created(
+    owner_id: felt,
+    game_id : felt,
+    state : felt
+):
+end
+
+@event
+func game_evolved(
+    user_id : felt,
+    game_id : felt,
+    generation : felt,
+    state : felt
+):
+end
+
+@event
+func cell_revived(
+    user_id : felt,
+    generation : felt,
+    cell_index : felt,
+    state : felt
+):
+end
+
+##########
+
+func pay{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt, 
+        credit_requirement : felt
+    ):
+    ERC20._burn(user, Uint256(credit_requirement, 0))
+    return ()
+end
+
+func reward_user{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt
+    ):
+    ERC20._mint(user, Uint256(1, 0))
+    return ()
+end
+
+func ensure_user{
+        syscall_ptr : felt*
+    }(
+    ) -> (
+        caller : felt
+    ):
+    let (caller) = get_caller_address()
+    with_attr error_message(
+        "User not authenticated"
+    ):
+        assert_not_zero(caller)
+    end
+    return (caller)
+end
+
+func assert_valid_new_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game : felt
+    ):
+    alloc_locals
+    let (local game_info) = stored_game.read(
+        game_id=game, generation=0)
+    let game_id = game_info[1]
+
+    with_attr error_message(
+        "Game with genesis state {game} already exists"
+    ):
+        assert_not_equal(game_id, game)
```

Comment:
Makes sense, although I think we should actually assert the saved game is zero (if that's the case, we can create new game; otherwise game already exists and we should error out). Added in newest commit.

Reply:
My bad, I meant assuming that saved game is 0 :)